### PR TITLE
Master merge into wasm integeration branch

### PR DESCRIPTION
**Pull request by umairkhannn**

### CODE REVIEW

File: lib/saito/ui/saito-crypto/overlays/deposit.js [hunk 1]
```diff
@@ -4,44 +4,64 @@ const SaitoOverlay = require("./../../saito-overlay/saito-overlay");
...
+	this.address = this.address.split("|")[0];
+      }
       this.render();
     });
   }
 
   render() {
-    this.overlay.show(DepositTemplate(this.app, this.mod, this));
+    this.overlay.show(DepositTemplate(this.app, this.mod, this));  
+    this.renderCrypto();
```
Comment (SankaD) [inline lib/saito/ui/saito-crypto/overlays/deposit.js:27]: shouldn't this be await and async render()

File: lib/saito/ui/saito-header/saito-header.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ const MyUserMenu = require("../../ui/modals/my-user-menu/my-user-menu.js");
 //
 class SaitoHeader extends UIModTemplate {
   constructor(app, mod) {
-    super(app, mod);
+    super(app);
```
Comment (SankaD) [inline lib/saito/ui/saito-header/saito-header.js:17]: intentional ?

File: lib/saito/ui/saito-header/saito-header.js [hunk 3]
```diff
@@ -83,7 +85,7 @@ class SaitoHeader extends UIModTemplate {
   }
 
   async render() {
-    let this_header = this;
+    this_header = this;
```
Comment (SankaD) [inline lib/saito/ui/saito-header/saito-header.js:88]: !!!

File: lib/saito/ui/saito-header/saito-header.js [hunk 4]
```diff
@@ -160,6 +162,9 @@ class SaitoHeader extends UIModTemplate {
     await this.app.modules.renderInto(".saito-header");
     await this.app.modules.renderInto(".saito-header-themes");
 
+    this.app.modules.renderInto(".saito-header");
```
Comment (SankaD) [inline lib/saito/ui/saito-header/saito-header.js:165]: duplicated? line 162

File: lib/saito/ui/saito-header/saito-header.js [hunk 5]
```diff
@@ -306,9 +315,9 @@ class SaitoHeader extends UIModTemplate {
       };
     } catch (err) {}
 
-    document.querySelector(".pubkey-containter").onclick = async (e) => {
+    document.querySelector(".pubkey-containter").onclick = (e) => {
       let public_key = document.getElementById("profile-public-key").innerHTML;
-      await navigator.clipboard.writeText(public_key);
+      navigator.clipboard.writeText(public_key);
```
Comment (SankaD) [inline lib/saito/ui/saito-header/saito-header.js:320]: async awiat?

File: lib/saito/ui/saito-header/saito-header.js [hunk 6]
```diff
@@ -394,8 +415,12 @@ class SaitoHeader extends UIModTemplate {
...
         }
       } else {
-        if (document.querySelector("#profile-public-key").innerHTML != this.publicKey) {
-          document.querySelector("#profile-public-key").innerHTML = this.publicKey;
+        if (
+          document.querySelector("#profile-public-key").innerHTML !=
+          this.publicKey
+        ) {
+          document.querySelector("#profile-public-key").innerHTML =
+            this.app.wallet.returnPublicKey();
```
Comment (SankaD) [inline lib/saito/ui/saito-header/saito-header.js:423]: await getPublicKey() ? or this.publicKey

File: lib/saito/ui/saito-profile/saito-profile.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        this.attachEvents();
+    }
+
+    attachEvents(){
+
+    }
+
+    addProfileMenu() {
+        this_profile = this;
+        let mods = this.app.modules.respondTo("saito-profile-menu", {publickey: this.publickey});
```
Comment (SankaD) [inline lib/saito/ui/saito-profile/saito-profile.js:34]: await

File: lib/saito/ui/saito-profile/saito-profile.template.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        <div class="saito-profile-icons">
+          <!--<i class="fas fa-gamepad"></i>    
+          <i class="far fa-id-card"></i>
+          <i class="fas fa-video"></i>
+          <i class="fa fa-money-bill-wave"></i>-->
+    `;
+
+    if (
+      app.keychain.hasPublicKey(profile.publickey) &&
+      profile.publickey !== app.wallet.returnPublicKey()
```
Comment (SankaD) [inline lib/saito/ui/saito-profile/saito-profile.template.js:23]: get

File: lib/templates/gametabletemplate.js [hunk 9]
```diff
@@ -9,111 +9,69 @@
...
-    this.scoreFrame = ""; //We are hardcoding scoreboard, but this is a hook to let the game module include information
     this.statistical_unit = "hand";
+    
   }
 
-  async initialize(app) {
-    await super.initialize(app);
-  }
+  render(app) {
+    super.render(app);
```
Comment (SankaD) [inline lib/templates/gametabletemplate.js:31]: await

File: lib/templates/gametabletemplate.js [hunk 10]
```diff
@@ -9,111 +9,69 @@
...
-                game_self.controller.bind(game_self)
-              );
-            } else {
-              game_self.willleave = false;
-              game_self.scoreboard.update(
-                this.scoreFrame + `<div id="leave" class="table_ctrl">LEAVE</div>`,
-                game_self.controller.bind(game_self)
-              );
+    if (this.game.players.length < this.maxPlayers && this.game.options["open-table"]) {
+      if (!this.game.eliminated || !this.game.eliminated.includes(app.wallet.returnPublicKey())) {
```
Comment (SankaD) [inline lib/templates/gametabletemplate.js:58]: get

File: lib/templates/gametabletemplate.js [hunk 11]
```diff
@@ -122,7 +80,7 @@ class GameTableTemplate extends GameTemplate {
       module: this.name,
       game_id: this.game.id,
       request: request,
-      my_key: this.publicKey,
+      my_key: this.publickey,
```
Comment (SankaD) [inline lib/templates/gametabletemplate.js:83]: check why this is key instead of Key

File: lib/templates/gametabletemplate.js [hunk 12]
```diff
@@ -231,14 +181,14 @@ class GameTableTemplate extends GameTemplate {
...
-              `${app.keychain.returnUsername(gametxmsg.my_key)} changed their mind`,
-              2500
-            );
+            siteMessage(`${app.keychain.returnUsername(gametxmsg.my_key)} changed their mind`, 2500);
           }
         }
       }
     }
+
+    super.handlePeerTransaction(app, tx, peer, mycallback);
```
Comment (SankaD) [inline lib/templates/gametabletemplate.js:190]: await?

File: lib/templates/gametabletemplate.js [hunk 13]
```diff
@@ -249,8 +199,8 @@ class GameTableTemplate extends GameTemplate {
     //To add a player after the game started,
     // need to assign this.game.player
     // add key
-    if (this.publicKey === address) {
-      this.game.live = 1;
+    if (this.publickey === address) {
```
Comment (SankaD) [inline lib/templates/gametabletemplate.js:202]: k vs K

File: lib/templates/gametabletemplate.js [hunk 14]
```diff
@@ -320,26 +268,28 @@ class GameTableTemplate extends GameTemplate {
           this.addPlayerToState(pkey);
           await this.addPlayerLate(pkey);
           this.updateLog(
-            `${this.app.keychain.returnUsername(pkey)} joins the table as Player ${
-              this.game.players.length
-            }`
+            `${this.app.keychain.returnUsername(pkey)} joins the table as Player ${this.game.players.length}`
           );
-          if (pkey === this.publicKey) {
+          if (pkey === this.publickey) {
```
Comment (SankaD) [inline lib/templates/gametabletemplate.js:273]: k vs K

File: lib/templates/gametemplate-moves.js [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+    // HALTED  -- a safety catch for future moves
+    ////////////
+    if (this.game.halted == 1 || this.gaming_active == 1 || this.initialize_game_run == 0) {
+      console.info(`Save as future move because halted (${this.game.halted}) or active (${this.gaming_active})`);
+      this.addFutureMove(gametx);
+      return;
+    }
+
+    //Update player's step value
+    this.game.step.players[gametx.transaction.from[0].add] = gametxmsg.step.game;
```
Comment (SankaD) [inline lib/templates/gametemplate-moves.js:111]: add vs publicKey

File: lib/templates/gametemplate-moves.js [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+
+  addFutureMove(gametx) {
+    if (!this.game.future) {
+      this.game.future = [];
+    }
+
+    if (!this.game.future.includes(JSON.stringify(gametx.transaction))) {
+      
+      console.log("Future TX:" , JSON.parse(JSON.stringify(gametx.returnMessage())));
+      this.game.future.push(JSON.stringify(gametx.transaction));
```
Comment (SankaD) [inline lib/templates/gametemplate-moves.js:164]: check fields in transaction

File: lib/templates/gametemplate-moves.js [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+    if (this.game.halted == 1 || this.initialize_game_run == 0) {
+      console.info(`Unable to process future moves now because halted (${this.game.halted}) or gamefeeder not initialized (${this.initialize_game_run})`);
+      return -1;
+    }
+
+    //console.info(this.game.future.length + " FUTURE MOVES at step --> " + this.game.step.game);
+
+    //Search all future moves for the next one
+    for (let i = 0; i < this.game.future.length; i++) {
+      let ftx = new saito.default.transaction(JSON.parse(this.game.future[i]));
```
Comment (SankaD) [inline lib/templates/gametemplate-moves.js:196]: constructor usage is different now

File: lib/templates/gametemplate-moves.js [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+  }
+
+  /**
+   * Check what moves are pending in my wallet and if any of them are future moves
+   * (I sent, but haven't received and processed) return 1
+   */
+  hasGameMovePending() {
+    for (let i = 0; i < this.app.wallet.wallet.pending.length; i++) {
+      let tx = new saito.default.transaction();
+      tx.deserialize_from_web(this.app, this.app.wallet.wallet.pending[i]);
```
Comment (SankaD) [inline lib/templates/gametemplate-moves.js:281]: wallet.wallet is not correct now

**Review by SankaD — CHANGES_REQUESTED** [review]

File: lib/saito/modules.ts [hunk 19]
```diff
@@ -220,9 +220,10 @@ class Mods {
     // .. and setup active module
     //
     if (this.app.BROWSER) {
+      await this.app.modules.render();
```
[the remark this answers is not in the document]
Comment (umairkhannn) [reply lib/saito/modules.ts:223]: no significance; this change is from master pull.

File: lib/templates/gametemplate-moves.js (hunk 16 above)
Comment (umairkhannn) [reply lib/templates/gametemplate-moves.js:169]: what is to check here? Sorry didn't understand.

File: cli.ts [hunk 20]
```diff
@@ -18,9 +16,6 @@ async function initCLI() {
   // eslint-disable-next-line @typescript-eslint/ban-ts-comment
   // @ts-ignore
   app.storage = new StorageCore(app);
-  // const blake3 = require("blake3-wasm");
-  // const blake3 = await import ("blake3");
-  await hashLoader(app);
```
Comment (SankaD) [inline cli.ts:23]: why is this line removed? this is needed for hashing

File: lib/saito/ui/game-pokerbox/game-playerbox.js [hunk 21]
```diff
@@ -0,0 +1,560 @@
...
+    for (let i = 0; i < this.app.modules.mods.length; i++) {
+      if (this.app.modules.mods[i].slug === "chat") {
+        chatmod = this.app.modules.mods[i];
+      }
+    }
+
+    for (let player = 1; player <= this.mod.game.players.length; player++) {
+      let boxId = this.playerBox(player);
+      $(`#player-box-head-${boxId}`).off();
+      if (pb_self.mod.game.players[player - 1] !== pb_self.app.wallet.returnPublicKey()) {
```
Comment (SankaD) [inline lib/saito/ui/game-pokerbox/game-playerbox.js:74]: getPublicKey ?

File: lib/saito/ui/game-pokerbox/game-playerbox.js [hunk 22]
```diff
@@ -0,0 +1,560 @@
...
+      if (document.getElementById("player-box-1")){
+        document.querySelector("#player-box-1").after(oBox); //Put new wrapper just after the player box  
+      }else{
+        document.body.append(oBox);
+      }
+      
+    }
+
+    let opponents = this.returnPlayersBoxArray();
+    if (!this.mod.game.players.includes(this.app.wallet.returnPublicKey())){
```
Comment (SankaD) [inline lib/saito/ui/game-pokerbox/game-playerbox.js:161]: get

File: lib/saito/ui/game-pokerbox/game-playerbox.js [hunk 23]
```diff
@@ -0,0 +1,560 @@
...
+   * @param {int} pnum - player number, e.g. {1, 2, ... # of players}
+   * Converts the player number to a "seat position" This player is always 1, unless you render with game.seats
+   */
+  playerBox(pnum) {
+    //For attempts to access playerbox before rendered
+    if (!this.mod) {
+      return 1;
+    }
+
+    if (!this.mod.game.players.includes(this.app.wallet.returnPublicKey())) {
```
Comment (SankaD) [inline lib/saito/ui/game-pokerbox/game-playerbox.js:197]: get

File: lib/saito/ui/saito-crypto/overlays/deposit.js [hunk 24]
```diff
@@ -4,44 +4,64 @@ const SaitoOverlay = require("./../../saito-overlay/saito-overlay");
...
     this.app = app;
     this.mod = mod;
-    this.container = container;
-    this.overlay = new SaitoOverlay(this.app, this.mod);
-    this.address = "";
-    this.confs = "";
-    this.ticker = "";
+    this.overlay = new SaitoOverlay(app, mod, false);
+    this.ticker = "SAITO";
+    this.address = this.app.wallet.returnPublicKey();
```
Comment (SankaD) [inline lib/saito/ui/saito-crypto/overlays/deposit.js:12]: get await

File: lib/saito/ui/saito-profile/saito-profile.template.js [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+        <div class="saito-profile-icons">
+          <!--<i class="fas fa-gamepad"></i>    
+          <i class="far fa-id-card"></i>
+          <i class="fas fa-video"></i>
+          <i class="fa fa-money-bill-wave"></i>-->
+    `;
+
+    if (
+      app.keychain.hasPublicKey(profile.publickey) &&
+      profile.publickey !== app.wallet.getPublicKey()
```
Comment (SankaD) [inline lib/saito/ui/saito-profile/saito-profile.template.js:23]: await

File: lib/templates/gametabletemplate.js [hunk 26]
```diff
@@ -9,111 +9,69 @@
...
-                game_self.controller.bind(game_self)
-              );
-            } else {
-              game_self.willleave = false;
-              game_self.scoreboard.update(
-                this.scoreFrame + `<div id="leave" class="table_ctrl">LEAVE</div>`,
-                game_self.controller.bind(game_self)
-              );
+    if (this.game.players.length < this.maxPlayers && this.game.options["open-table"]) {
+      if (!this.game.eliminated || !this.game.eliminated.includes(app.wallet.getPublicKey())) {
```
Comment (SankaD) [inline lib/templates/gametabletemplate.js:58]: await

File: lib/templates/gametemplate-moves.js [hunk 27]
```diff
@@ -0,0 +1,306 @@
...
+    // HALTED  -- a safety catch for future moves
+    ////////////
+    if (this.game.halted == 1 || this.gaming_active == 1 || this.initialize_game_run == 0) {
+      console.info(`Save as future move because halted (${this.game.halted}) or active (${this.gaming_active})`);
+      this.addFutureMove(gametx);
+      return;
+    }
+
+    //Update player's step value
+    this.game.step.players[gametx.transaction.from[0].publicKey] = gametxmsg.step.game;
```
Comment (SankaD) [inline lib/templates/gametemplate-moves.js:116]: tx doesn't have a transaction field now

File: lib/templates/gametemplate-moves.js (hunk 16 above)
Comment (SankaD) [reply lib/templates/gametemplate-moves.js:169]: Tx doesn't have a transaction field now. use toJson() perhaps?

File: lib/templates/gametemplate-moves.js [hunk 28]
```diff
@@ -0,0 +1,306 @@
...
+
+    //console.info(this.game.future.length + " FUTURE MOVES at step --> " + this.game.step.game);
+
+    //Search all future moves for the next one
+    for (let i = 0; i < this.game.future.length; i++) {
+      let ftx = new saito.default.transaction(JSON.parse(this.game.future[i]));
+      let ftxmsg = ftx.returnMessage();
+      //console.info(`FTMSG (${ftxmsg.step.game}): ` + JSON.stringify(ftxmsg.turn));
+
+      if (this.isUnprocessedMove(ftx.transaction.from[0].publicKey, ftxmsg)) {
```
Comment (SankaD) [inline lib/templates/gametemplate-moves.js:205]: tx.from[0]

File: lib/templates/gametemplate-moves.js [hunk 29]
```diff
@@ -0,0 +1,306 @@
...
+
+    if (this.app.options?.games) {
+      for (let i = 0; i < this.app.options.games.length; i++) {
+        if (this.app.options.games[i].id === this.game.id) {
+          this.app.options.games[i].future = this.game.future;
+        }
+      }
+    }
+
+    this.app.storage.saveOptions();
```
Comment (SankaD) [inline lib/templates/gametemplate-moves.js:252]: await

**Review by SankaD — CHANGES_REQUESTED** [review]

File: lib/templates/gametemplate-moves.js [hunk 30]
```diff
@@ -0,0 +1,306 @@
...
+   */
+  hasGameMovePending() {
+    for (let i = 0; i < this.app.wallet.wallet.pending.length; i++) {
+      let tx = new saito.default.transaction();
+      tx.deserialize_from_web(this.app, this.app.wallet.pending[i]);
+      let txmsg = tx.returnMessage();
+      if (txmsg && txmsg.module == this.name) {
+        if (txmsg.game_id === this.game?.id) {
+          if (txmsg?.step?.game) {
+            if (txmsg.step.game > this.game.step.players[tx.transaction.from[0].publicKey]) {
```
Comment (SankaD) [inline lib/templates/gametemplate-moves.js:291]: tx.transaction

File: lib/templates/gametemplate-players.js [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+      }
+    }
+
+    this.game.players.push(address);
+
+    if (!this.game.accepted.includes(address)) {
+      this.game.accepted.push(address);
+    }
+
+    if (this.app.wallet.returnPublicKey() !== address) {
```
Comment (SankaD) [inline lib/templates/gametemplate-players.js:58]: get

File: lib/templates/oneplayergametemplate.js [hunk 32]
```diff
@@ -1,242 +1,266 @@
...
+  async render(app){
+
+    if (this.game.player == 0){
+      let c = await sconfirm("This is not a valid game! Return to Arcade");
+      if (c) { 
+        window.location = "/arcade";
+      }
+    }
+
+    super.render(app);
```
Comment (SankaD) [inline lib/templates/oneplayergametemplate.js:106]: await?

**Review by SankaD — CHANGES_REQUESTED** [review]
